swarm/pss/snapshot_test.go

```diff
@@ -0,0 +1,433 @@
+	msgC         chan handlerNotification // message receipt notification to main sim thread
+	msgs         [][]byte                 // recipient addresses of messages
+)
+
+func resetTestVariables() {
+	handlerDone = false
+	msgsToReceive = 0
+	maxMessages = 0
+	msgCnt = 0
+	msgs = nil
+	sim = nil
+
+	kademlias = make(map[enode.ID]*network.Kademlia)
+	nodeAddrs = make(map[enode.ID][]byte)
+	recipients = make(map[int][]enode.ID)
+	allowed = make(map[int][]enode.ID)
+	expectedMsgs = make(map[enode.ID][]uint64)
+	allowedMsgs = make(map[enode.ID][]uint64)
+	senders = make(map[int]enode.ID)
+	handlerC = make(chan handlerNotification)
+	doneC = make(chan struct{})
+	errC = make(chan error)
+	msgC = make(chan handlerNotification)
+}
+
+func isDone() bool {
+	mu.Lock()
+	defer mu.Unlock()
+	return handlerDone
+}
+
+func setDone() {
+	mu.Lock()
+	defer mu.Unlock()
+	handlerDone = true
+}
+
+func getCmdParams(t *testing.T) (int, int) {
+	args := strings.Split(t.Name(), "/")
+	msgCount, err := strconv.ParseInt(args[1], 10, 16)
+	if err != nil {
+		t.Fatal(err)
+	}
+	nodeCount, err := strconv.ParseInt(args[2], 10, 16)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return int(msgCount), int(nodeCount)
+}
+
+func readSnapshot(t *testing.T, nodeCount int) simulations.Snapshot {
+	f, err := os.Open(fmt.Sprintf("testdata/snapshot_%d.json", nodeCount))
+	if err != nil {
+		t.Fatal(err)
+	}
+	jsonbyte, err := ioutil.ReadAll(f)
+	if err != nil {
+		t.Fatal(err)
+	}
+	var snap simulations.Snapshot
+	err = json.Unmarshal(jsonbyte, &snap)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return snap
+}
+
+func assingTestVariables(sim *simulation.Simulation, msgCount int) {
+	log.Debug("-------------------------------------------------------------------------")
+	for _, nodeId := range sim.NodeIDs() {
+		nodeAddrs[nodeId] = nodeIDToAddr(nodeId)
+	}
+
+	for i := 0; i < int(msgCount); i++ {
+		msgAddr := pot.RandomAddress() // we choose message addresses randomly
+		msgs = append(msgs, msgAddr.Bytes())
+		smallestPo := 256
+		var targets []enode.ID
+		var prev int
+
+		// loop through all nodes and add the message to recipient indices
+		for _, nod := range sim.Net.GetNodes() {
+			po, _ := pof(msgs[i], nodeAddrs[nod.ID()], 0)
+			depth := kademlias[nod.ID()].NeighbourhoodDepth()
+
+			// only nodes with closest IDs (wrt msg) will receive the msg
+			if po > prev {
+				prev = po
+				targets = nil
+				targets = append(targets, nod.ID())
+			} else if po == prev {
+				targets = append(targets, nod.ID())
+			}
+
+			if po >= depth {
+				maxMessages++
+				allowed[i] = append(allowed[i], nod.ID())
+				allowedMsgs[nod.ID()] = append(allowedMsgs[nod.ID()], uint64(i))
+			}
+
+			// a node with the smallest PO (wrt msg) will be the sender
+			if po < smallestPo {
+				smallestPo = po
+				senders[i] = nod.ID()
+			}
+		}
+
+		msgsToReceive += len(targets)
+		for _, id := range targets {
+			recipients[i] = append(recipients[i], id)
```

Comment:
why can't you assign directly to `recipients[]` above? (if you implement a struct to hold the state this will be a single method call to update both arrays).

Reply:
because in some cases targets might be reset. please see this line 157:
```
targets = nil
```